File: scripts/Teleop.py
#!/usr/bin/env python


"""
0: left/right(1 to -1)
1: backward/forward (-1 to 1)
"""

# ROS Imports
import rospy
from geometry_msgs.msg import Twist
from sensor_msgs.msg import Joy

sign = lambda x: (1, -1)[x < 0]

class Teleop:
    def __init__(self):
        rospy.init_node('teleop')

        self.joy_sub = rospy.Subscriber("/joy", Joy, self.joyCB)
        self.twist_pub = rospy.Publisher("/cmd_vel", Twist, queue_size=10)

    def joyCB(self, msg):
        joy_lin = msg.axes[1] * 0.2 # scale by max linear velocity (0.5 max)
        joy_ang = sign(joy_lin) * msg.axes[0] * 1   # scale by max angular velocity (4 max)

        cmd = Twist()
        cmd.linear.x = joy_lin
        cmd.angular.z = joy_ang
        self.twist_pub.publish(cmd)

if __name__ == "__main__":
    teleop = Teleop()
    rospy.spin()

File: test/Teleop/Teleop.cpp
//
// Created by amy on 6/20/20.
//

#include "Teleop.h"


RoombaTeleop::RoombaTeleop():
        linear_(1),
        angular_(2)
{

    nh_.param("axis_linear", linear_, linear_);
    nh_.param("axis_angular", angular_, angular_);
    nh_.param("scale_angular", a_scale_, a_scale_);
    nh_.param("scale_linear", l_scale_, l_scale_);


    vel_pub_ = nh_.advertise<geometry_msgs::Twist>("/cmd_vel", 1);

    joy_sub_ = nh_.subscribe<sensor_msgs::Joy>("joy", 10, &RoombaTeleop::joyCallback, this);

}

void RoombaTeleop::joyCallback(const sensor_msgs::Joy::ConstPtr& joy)
{
    geometry_msgs::Twist twist;
    twist.angular.z = a_scale_*joy->axes[angular_];
    twist.linear.x = l_scale_*joy->axes[linear_];
    vel_pub_.publish(twist);
}


int main(int argc, char** argv)
{
    ros::init(argc, argv, "RoombaTeleop");
    RoombaTeleop roomba_teleop;

    ros::spin();
}